### DISCUSSION TITLE
Walkthrough for complete newbies? Trouble building for STM32F103C8

### DISCUSSION THREAD

**Discussion by avizienis** [terjeio/grblHAL#154] [Q&A] [answered]
Hi, I just spent full day tinkering with IDEs and am now completely lost.
My goal is to build a grblHAL for the blue-pill STM32 module for 3 axis and with maximum compatibility with hopes to run it with bCNC (which I like but it's very picky when it comes to G CODE format. So far none of the STM32 forks I tried seem to work)
So I did all the compile time configuration steps - copied source files to STM32F1xx folder, edited config files etc. Then loaded into Atollic and nothing happened, just a single line "nothing to build for project..."
Then I stumbled randomly through project settings and found that there is something wrong with the toolchain:
Orphaned toolchain com.st.stm32cube.ide.mcu.gnu.managedbuild.toolchain.exe.release.2114697922 (MCU ARM GCC)
And there are no other compatible toolchains installed, although I had just installed the newest GNU Arm Embedded Toolchain on top of Atollic. I also tried all the non-compatible toolchains, all of them gave various errors.
So I figured I must have the wrong software, so I wiped out Atollic and installed Eclipse. Then added "CDT main" or something like that through Help>install new software. And there is no progress, still get this in the console:

**** Build of configuration Release for project GRBL Driver STM32F103C8 ****

Nothing to build for project GRBL Driver STM32F103C8

And this in project settings:
![image](https://user-images.githubusercontent.com/7872840/103002580-d1c81280-4537-11eb-8e88-ef9abbceeb21.png)

Now I know for advanced users this sounds like a "go RTFM" type of situation but all previous project I've tried just worked out of the box. I used to work with Visual Studio C++ back in the day but the sheer number of configuration options in these modern packages is just overwhelming.

Any help appreciated, thanks!

**Comment by terjeio** [comment, marked as answer]
The driver is built with the STM32CubeIDE - see the [driver page](https://github.com/terjeio/grblHAL/tree/master/drivers/STM32F1xx). The project files are for that environment.

Configuring other toolchains may be possible but would require importing the project without the supplied project files (delete them?) and setting up the the project from scratch?

**Reply by avizienis** [reply]
Thank you very much, that did the trick. Compiled right away without problems!

**Reply by avizienis** [reply]
However bCNC didn't like it, locks up in an alarm state:
```
error:9

ok
$X
error:18
$X
error:18
$G
[GC:G0 G54 G17 G21 G90 G94 M5 M9 T0 F0 S0.]
ok
$X
error:18
```
Any obvious pointers to what did I do wrong? I compiled with compatibility level 10

**Reply by terjeio** [reply]
See the [test branch readme](https://github.com/terjeio/grblHAL/tree/test/README.md) for important info about NC switches.

**Comment by phil-barrett** [comment]
The control inputs default to NC (Normally Closed) sensors (switches and such).  You can either pull them low on your board or invert them in Grbl Settings ($5 and $14).  The specific ones you should deal with right now are Safety Door and EStop ($14=72 will do it).  You will probably want to also invert Hold and Start ($14=78).  Set limits ($5) when you turn on hard limits ($21)

**Reply by avizienis** [reply]
Thanks guys, that did the trick! Now it works with an older version of bCNC, but the newest seems to restart every couple of seconds:
`Grbl 1.1f ['$' for help]

[G54:10.000,0.000,0.000]
[G55:0.000,0.000,0.000]
[G56:0.000,0.000,0.000]
[G57:0.000,0.000,0.000]
[G58:0.000,0.000,0.000]
[G59:0.000,0.000,0.000]
[G28:0.000,0.000,0.000]
[G30:0.000,0.000,0.000]
[G59.3:0.000,0.000,0.000]
[G92:0.000,0.000,0.000]
[TLO:0.000]
[PRB:0.000,0.000,0.000:0]
ok
G90
ok
Run ended
2020-12-23 18:23:50.424531
Current: 2 [2]  Completed: 100% [0s Tot: 0s ]
`
But I guess that's a problem with the software, not grbl. Will try to ask in their forums

**Reply by terjeio** [reply]
There is a regression at play here: `[G59.3:0.000,0.000,0.000]` should not be output when compatibility mode is 10. Could this cause the restart?

I have just committed a fix to the [test branch](https://github.com/terjeio/grblHAL/tree/test) for this. There are some other changes in this as well that got committed a bit prematurely - but those should not cause any problems.

**Reply by avizienis** [reply]
Compiled the test branch, result is virtually identical, regardless of compatibility level. This is the terminal dump from bCNC:
```
Grbl 1.1f ['$' for help]

[G54:0.000,0.000,0.000]
[G55:0.000,0.000,0.000]
[G56:0.000,0.000,0.000]
[G57:0.000,0.000,0.000]
[G58:0.000,0.000,0.000]
[G59:0.000,0.000,0.000]
[G28:0.000,0.000,0.000]
[G30:0.000,0.000,0.000]
[G92:0.000,0.000,0.000]
[TLO:0.000]
[PRB:0.000,0.000,0.000:0]
ok
G90
ok
Run ended
2020-12-23 22:54:53.716068
Current: 2 [2]  Completed: 100% [0s Tot: 0s ]
```

**Reply by terjeio** [reply]
I cannot get bCNC to run because I do not have Tkinter in my python install. However, I have tested with four other senders and all works.

Are you able to debug? If so we can check if bCNC for some reason sends reset commands triggering this behaviour. Set breakpoints at line 705 and 712 to do that.

https://github.com/terjeio/grblHAL/blob/3870091f97d1a11ef668ee67be4eb90400bf178d/grbl/protocol.c#L704-L715

**Reply by avizienis** [reply]
Thanks for the tip, I might do that after the holidays. For now I think I'm just gonna stick to cncjs which seems to be better supported and during initial testing worked without any problems whatsoever.

**Reply by avizienis** [reply]
Update: it seems that bCNC does indeed work, even with compatibility zero. It was a $14 issue after all. Older bCNC version and cncjs simply don't care about the alarms.
bCNC does have hard-coded grbl configurator so that refuses to display anything.

**Comment by phil-barrett** [comment]
I just added a section to the wiki that might be worth reviewing. https://github.com/terjeio/grblHAL/wiki/First-Run-Grbl-Settings

Comments, suggestions, questions and criticisms are welcome.

**Comment by avizienis** [comment]
One more question, in the drivers page it is stated that Inverted spindle PWM is available for STM32F1 but not for STM32F4, which is strange in itself but even more, I can't find any references to this in the config files (my_machine, maps, config.h).
Where can I find this clause and is it really not supported on STM32F4 boards?

**Reply by terjeio** [reply]
It is supported but not documented in [the wiki](https://github.com/terjeio/grblHAL/wiki/Additional-or-extended-settings) as it should, I'll fix that and add options to config.h.

Inversion _is_ settable by bit 2 for setting `$16`, e.g. `$16=4` will invert PWM only. If you use [ioSender](https://github.com/terjeio/Grbl-GCode-Sender) then this is available from the grbl settings tab.
